File: core/src/aws-entries.test.ts
/**
 * Regression tests for the "AWS SSO" / "AWS" report: two entries share
 * username and password but have different OTP URIs. The core layers (vault
 * model, encrypted store round trip, OTP canonicalization) must never merge or
 * deduplicate entries by username/password: entries are keyed by name only.
 */
import { describe, expect, it } from "vitest";
import { generateTotp, normalizeOtpInput, parseOtpauthUri } from "./otp.js";
import { loadStore, saveStore } from "./store.js";
import { addEntry, newVault, type Entry } from "./vault.js";

// Fixture secrets only — never real ones.
const SSO_OTP = "otpauth://totp/AWS%20SSO:alice?secret=JBSWY3DPEHPK3PXP&issuer=AWS%20SSO";
const AWS_OTP = "otpauth://totp/Amazon%20Web%20Services:alice@123456789012?secret=KRSXG5CTMVRXEZLUKN2XAZLSKNSWG4TFOQ&issuer=Amazon%20Web%20Services";

function awsEntry(otp: string): Entry {
  return {
    metadata: { username: "alice", password: "same-password", url: "https://console.aws.amazon.com", otp },
    created_at: "2026-09-20T10:00:00.000Z",
    updated_at: "2026-09-20T10:00:00.000Z",
  };
}

describe("two entries with the same username/password, different OTP", () => {
  it("the fixture OTPs really are different secrets producing different codes", async () => {
    expect(parseOtpauthUri(SSO_OTP).secret).not.toBe(parseOtpauthUri(AWS_OTP).secret);
    const t = Date.UTC(2026, 8, 24, 12, 0, 0);
    expect((await generateTotp(SSO_OTP, t)).code).not.toBe((await generateTotp(AWS_OTP, t)).code);
  });

  it("the vault model keys entries by name only: both are kept", () => {
    const v = newVault();
    addEntry(v, "AWS SSO", awsEntry(SSO_OTP));
    addEntry(v, "AWS", awsEntry(AWS_OTP));
    expect(Object.keys(v.entries).sort()).toEqual(["AWS", "AWS SSO"]);
    expect(v.entries["AWS"].metadata.otp).toBe(AWS_OTP);
    expect(v.entries["AWS SSO"].metadata.otp).toBe(SSO_OTP);
  });

  it("survives the encrypted store round trip unchanged", async () => {
    const v = newVault();
    addEntry(v, "AWS SSO", awsEntry(SSO_OTP));
    addEntry(v, "AWS", awsEntry(AWS_OTP));
    const back = await loadStore(await saveStore(v, "fixture-password"), "fixture-password");
    expect(back).toEqual(v);
  }, 30_000); // real Argon2id

  it("OTP canonicalization keeps each entry's own secret", async () => {
    const sso = await normalizeOtpInput(SSO_OTP, "AWS SSO");
    const aws = await normalizeOtpInput(AWS_OTP, "AWS");
    expect(parseOtpauthUri(sso).secret).toBe(parseOtpauthUri(SSO_OTP).secret);
    expect(parseOtpauthUri(aws).secret).toBe(parseOtpauthUri(AWS_OTP).secret);
  });
});

File: cli/cmd/localpass/aws_entries_test.go
package main

// Regression test for the "AWS SSO" / "AWS" report: two entries sharing
// username and password but with different OTPs must round-trip through sync
// as two separate entries, each with its own OTP. Uses temp stores and an
// in-process S3 stand-in only.

import (
	"io"
	"net/http"
	"net/http/httptest"
	"path/filepath"
	"reflect"
	"sync"
	"testing"

	"github.com/nawarian/localpass/cli/internal/config"
	"github.com/nawarian/localpass/cli/internal/store"
)

const (
	ssoOTP = "otpauth://totp/AWS%20SSO:alice?secret=JBSWY3DPEHPK3PXP&issuer=AWS%20SSO"
	awsOTP = "otpauth://totp/Amazon%20Web%20Services:alice@123456789012?secret=KRSXG5CTMVRXEZLUKN2XAZLSKNSWG4TFOQ&issuer=Amazon%20Web%20Services"
)

// plainS3 stores one object per path with unconditional PUTs, like S3 does
// for the requests main sends.
func plainS3(t *testing.T) *httptest.Server {
	var mu sync.Mutex
	objects := map[string][]byte{}
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		mu.Lock()
		defer mu.Unlock()
		switch r.Method {
		case http.MethodPut:
			body, _ := io.ReadAll(r.Body)
			objects[r.URL.Path] = body
		case http.MethodGet, http.MethodHead:
			data, ok := objects[r.URL.Path]
			if !ok {
				w.WriteHeader(http.StatusNotFound)
				return
			}
			if r.Method == http.MethodGet {
				w.Write(data)
			}
		}
	}))
	t.Cleanup(srv.Close)
	return srv
}

func reproEnv(t *testing.T) string {
	srv := plainS3(t)
	t.Setenv("AWS_ACCESS_KEY_ID", "AKID")
	t.Setenv("AWS_SECRET_ACCESS_KEY", "SECRET")
	t.Setenv("AWS_REGION", "us-east-1")
	t.Setenv("S3_ENDPOINT", srv.URL)
	cfg := filepath.Join(t.TempDir(), "config.json")
	config.SaveConfig(cfg, &config.Config{S3Bucket: "bucket", S3Key: "vault.enc", AutoSync: true})
	readPassword = func(fd int) ([]byte, error) { return []byte("pass"), nil }
	return cfg
}

func setEntry(store, cfg, key, otp string) {
	runSet([]string{key, "--store-path", store, "--config-path", cfg, "--password", "same-password",
		"--username", "alice", "--url", "", "--notes", "", "--meta", "otp=" + otp})
}

func entries(t *testing.T, path string) map[string]string {
	t.Helper()
	v, err := store.LoadStore(path, "pass")
	if err != nil {
		t.Fatalf("LoadStore: %v", err)
	}
	out := map[string]string{}
	for k, e := range v.Entries {
		out[k] = e.Metadata["otp"]
	}
	return out
}

func TestSyncKeepsEntriesWithSharedCredentialsApart(t *testing.T) {
	cfg := reproEnv(t)
	dev := filepath.Join(t.TempDir(), "store.json")
	runInit([]string{"--store-path", dev})
	setEntry(dev, cfg, "AWS SSO", ssoOTP)
	setEntry(dev, cfg, "AWS", awsOTP)

	fresh := filepath.Join(t.TempDir(), "store.json")
	runPull([]string{"--store-path", fresh, "--config-path", cfg, "--force"})
	want := map[string]string{"AWS SSO": ssoOTP, "AWS": awsOTP}
	if got := entries(t, fresh); !reflect.DeepEqual(got, want) {
		t.Fatalf("remote = %v, want %v", got, want)
	}
}
